### ISSUE TITLE
Scroll position is restored when not needed

### ISSUE THREAD

**Issue by catamphetamine** [taion/scroll-behavior#215]
I found a bug in scroll position restoration in `found` library.

The scroll position is stored in Session storage:

```
@@scroll\|/vg | [0,666.5] |  
```

where `/vg` is the page path.

If a user opens a page, then goes to the URL bar and enters another page URL, then goes to the URL bar again and enters the initial page URL — the page loads and the scroll is at where the user left it initially, not at the top of the page, because Session storage still has that entry. And the entry has no random-generated hash or anything.

**Comment by taion** [comment]
That's a good catch. The issue here is that the initial location doesn't get supplemented with a randomly-generated key, as we didn't want to trigger an immediate navigation action on page load.

Not quite sure how to deal with this one.

That said, this isn't specific to Found. The same thing will happen on e.g. most sites that use anything like this logic for determining page key, including Gatsby sites. You can see this behavior on e.g. https://reactjs.org.

cc @KyleAMathews @jquense – older versions of React Router actually _would_ immediately redirect to a version of the page with a random `key` set on load, but I think users tended to not like that behavior much. Not sure what the best way to deal with it is. I don't think there's a way to deal with it inside this library, though, but it is a weird bug.

**Comment by catamphetamine** [comment]
@taion Yeah, I recall that in `react-router@3` every location had its own random-generated `key` and I didn't like the idea of having it.

For the current situation, I was thinking about clearing all `@@scroll|` keys from Session storage on page load:

```js
window.onload = function onPageLoad() {
  forEach((key) => {
    if (key.indexOf('@@scroll|') === 0) {
      sessionStorage.removeItem(key)
    }
  }
}

function forEach(func) {
  let i = 0
  while (i < sessionStorage.length) {
    const key = sessionStorage.key(i)
    func(key)
    i++
  }
}
```

**Comment by taion** [comment]
That's not quite the same, though. To emulate native browser scroll behavior, you _do_ want to preserve scroll position if someone reloads the page.

**Comment by catamphetamine** [comment]
@taion Aaah, I see.

**Comment by taion** [comment]
And to be clear, we do still attach keys on navigation to locations _after_ the first. But we don't do this for the first location because it would require triggering a history action to get it stamped on.

**Comment by hedgepigdaniel** [comment]
Whats wrong with storing a random key on each location entry?

That makes perfect sense to me - a key to identify each one so that even if the URL is the same, there is something to distinguish 2 separate location entries that should not share the same scroll position.

Hard to see how this library could work properly if the locations provided to it are not unique.

**Comment by taion** [comment]
Farce adds unique keys to all locations that _it_ generates.

However, for the initial page load, I'm not aware of a way to update the location state. Old versions of `history` used to deal with this by immediately calling `replaceState`, but that led to weird bugs.

**Comment by hedgepigdaniel** [comment]
What kind of bugs did it cause?

**Comment by taion** [comment]
I honestly can't remember any more. It was enough that they took it out in history v4, though.

**Comment by catamphetamine** [comment]
> I honestly can't remember any more. It was enough that they took it out in history v4, though.

I did a quck search in the `history` repo's issues and found only one instance of such bug which happened in IE 11: https://github.com/remix-run/history/issues/220

That small bug was enough for one of the repo authors to discard the `history.replaceState()` workaround on page load.

The author specifically added it as a feature previously in https://github.com/remix-run/history/issues/9 but after the IE 11 bug report he reconsidered and denied any similar feature requests: https://github.com/remix-run/history/issues/502

My personal opinion: since I've already forked `scroll-behavior` to apply some minor unrelated adjustments, I might as well reinstate the `history.replaceState()` workaround on page load. IE 11 isn't widely used nowadays anyway. And `document.referer` being empty doesn't break anything besides Google Analytics which is a very insignificant inconvenience that can totally be neglected.

If anyone's interested, the workaround will be available in `navigation-stack` package:
https://www.npmjs.com/package/navigation-stack
